books/books/middleware.py

```diff
@@ -0,0 +1,33 @@
+import random
+from scrapy.exceptions import IgnoreRequest
+
+
+class RandomUserAgent(object):
+    """Randomly rotate user agents based on a list of predefined ones"""
+
+    def __init__(self, agents):
+        self.agents = agents
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler.settings.getlist('USER_AGENTS'))
+
+    def process_request(self, request, spider):
+        request.headers.setdefault('User-Agent', random.choice(self.agents))
+
+
+class ErrorMonkeyMiddleware(object):
```

Comment:
It is being compiled (since .pyc file is generated).
About it's use : I'll check and inform.